### DISCUSSION TITLE
How to set maximum memory for Vercel serverless functions? Python runtime.

### DISCUSSION THREAD

**Discussion by mhmdibrahimm** [vercel/community#5861] [Help]
### Summary

I coudln't get it working by any means, and docs are poorly written for this matter. I couldn't find any resource on how to this for python. 
I want the maximum size to be 128MB, now it's 1024MB and I couldn't change it.

### Steps to Reproduce

My `vercel.json`:
```json
{
	"builds":[
		{
			"src":"main.py",
			"use":"@vercel/python"
		}
	],
	"routes":[
		{
			"src":"/(.*)",
			"dest":"main.py"
		}

	]
}
```
My flask app is in main, variable name is app
file structure:
markethunter/
┣ lib/
┣ README.md
┣ main.py
┣ requirements.txt
┗ vercel.json

**Comment by versecafe** [comment]
@iamEvara https://vercel.com/docs/functions/configuring-functions/memory It should be completely controllable from the `vercel.json`

```json
{
  "functions": {
    "route/to/function": {
      "memory": 124
    }
  }
}
```

**Reply by mhmdibrahimm** [reply]
what's "route/to/function" in my case? I tried this:
```json
{
	"functions": {
	  "main": {
		"memory": 128
	  }
	}
  }
```
also tried having `main.py` instead of `main`, and I got this error on deployment:
```
Error: The pattern "main" defined in `functions` doesn't match any Serverless Functions inside the `api` directory.
```
My flask app is in main, variable name is app
file structure:
markethunter/
┣ lib/
┣ README.md
┣ main.py
┣ requirements.txt
┗ vercel.json

main.py:

```py
from flask import Flask

app = Flask(__name__)

... main logic


if __name__ == "__main__":
    app.run(debug=False)
```

**Reply by versecafe** [reply]
@iamEvara Try using the API routes `/api/route`

**Comment by mhmdibrahimm** [comment]
I tried putting main.py under /api/ and building with the settings you've
provided but that didn't work I still get that error.

**Reply by versecafe** [reply]
@iamEvara I'll spin up a flask project today and try to get you a definitive answer on how to set function memory for Flask

**Reply by mhmdibrahimm** [reply]
Hey, any updates? We are paying for 1GB per 1 second when we're using 128mb. It's coming to a loss for us, could you please assist on that?

**Reply by flashlog** [reply]
@iamEvara Were you able to figure out the solution?

**Reply by versecafe** [reply]
@flashlog @iamEvara 
```json
{
  "rewrites": [{ "source": "/(.*)", "destination": "/api/index" }],
  "functions": {
    "api/index.py": {
      "memory": 128
    }
  }
}
```

**Reply by versecafe** [reply]
Do note that you cannot set under 128mb of memory or you'll get a 502 Bad Gateway error from Vercel, and you can test locally by using `vercel dev` to make debugging easier

<img width="881" alt="image" src="https://github.com/vercel/community/assets/147033096/aa9f6df9-4929-4f9d-885e-214828b95b12">

<img width="868" alt="image" src="https://github.com/vercel/community/assets/147033096/44095ff6-3cc1-4283-a76f-7bcc4d5d0a82">

**Reply by flashlog** [reply]
@versecafe Thanks, I was trying to increase the maxDuration which wasn't working if I had `main.py` at the root so I had to refactor the code according to the examples. (I'm using FastAPI)

Here's my vercel json config:
```
{
    "routes": [
        {
            "src": "/(.*)",
            "dest": "api/main.py"
        }
    ],
    "functions": {
        "api/main.py": {
            "maxDuration": 300
        }
    }
}
```
My updated directory:
![image](https://github.com/vercel/community/assets/64234223/c8ff8a03-2166-4392-932e-c1b2f4742e2b)

Now the issue is that Vercel keeps on installing the Python dependencies again and again (stuck)

![image](https://github.com/vercel/community/assets/64234223/3b1e8133-6655-4d34-baf7-d3da1f3800c1)

**Reply by versecafe** [reply]
@flashlog try forcing to 3.9 in the pipfile, the python runtime handling can be a bit less stable since fewer people use it.

**Reply by flashlog** [reply]
@versecafe I'll try but it worked before refactoring and moving the `main.py` file inside the `api` folder. I was using the following config:
```
{
    "builds": [
        {
            "src": "main.py",
            "use": "@vercel/python"
        }
    ],
    "routes": [
        {
            "src": "/(.*)",
            "dest": "main.py"
        }
    ]
}
```

I tried adding "maxDuration" inside `builds` but it didn't work:
```
{
    "builds": [
        {
            "src": "main.py",
            "use": "@vercel/python",
            "config": {
                "maxDuration": 300,
                "runtime": "python3.9"
            }
        }
    ],
    "routes": [
        {
            "src": "/(.*)",
            "dest": "main.py"
        }
    ]
}
```

![image](https://github.com/vercel/community/assets/64234223/781182c7-02f5-4f0b-89ee-f8c26dd4fc15)

The python runtime has alot of issues.

**Reply by flashlog** [reply]
@versecafe Still the same behavior, do you know what could be causing this?

![image](https://github.com/vercel/community/assets/64234223/17e6c810-a08e-468e-9331-b7cbbd392c5b)

**Reply by versecafe** [reply]
No idea, and for some reason existing python projects on Vercel now throw
<img width="1298" alt="image" src="https://github.com/vercel/community/assets/147033096/68636f25-bf78-45c7-b857-d2e9ed2e67b4">

This may actually be an issue on Vercel's side
